File: Advent2022/Day4/Day4.py
import aoc_utils as au

Area = tuple[int, int]


def contains(first: Area, second: Area) -> bool:
    return (first[0] <= second[0] and first[-1] >= second[-1]) or (
        second[0] <= first[0] and second[-1] >= first[-1]
    )


def test_one(input: list[tuple[Area, Area]]) -> int:
    return len([pair for pair in input if contains(*pair)])


def part_one(input: list[tuple[Area, Area]]) -> int:
    return len([pair for pair in input if contains(*pair)])


def overlap(first: Area, second: Area) -> bool:
    return (second[0] <= first[0] <= second[-1]) or (first[0] <= second[0] <= first[-1])


def test_two(input: list[tuple[Area, Area]]) -> int:
    return len([pair for pair in input if overlap(*pair)])


def part_two(input: list[tuple[Area, Area]]) -> int:
    return len([pair for pair in input if overlap(*pair)])


if __name__ == "__main__":
    pair_parser = au.RegexParser(
        [
            (
                "(\d*)-(\d*),(\d*)-(\d*)",
                lambda m: ((int(m[0]), int(m[1])), (int(m[2]), int(m[3]))),
            )
        ]
    )
    day = au.Day(
        2022,
        4,
        test_one,
        test_two,
        part_one,
        part_two,
        input=pair_parser,
        test_input=pair_parser,
    )

    day.run_all(run_tests=True)

File: Advent2022/Day4/Day4.cpp
#include "aoc_utils.h"

#include <string>
#include <regex>
#include <iostream>
#include <ostream>

using namespace std;

class AreaTuple
{

private:
    int first_start;
    int first_end;
    int second_start;
    int second_end;

public:
    AreaTuple(const string &s)
    {
        static const regex area_regex("(\\d*)-(\\d*),(\\d*)-(\\d*)");

        smatch base_match;

        if (std::regex_match(s, base_match, area_regex))
        {
            first_start = stoi(base_match[1].str());
            first_end = stoi(base_match[2].str());
            second_start = stoi(base_match[3].str());
            second_end = stoi(base_match[4].str());
        }
        else
            first_start = first_end = second_start = second_end = -1;
    }

    bool contains()
    {
        return (first_start <= second_start && first_end >= second_end) ||
               (second_start <= first_start && second_end >= first_end);
    }

    bool overlap()
    {
        return (second_start <= first_start && first_start <= second_end) ||
               (first_start <= second_start && second_start <= first_end);
    }
};

class Day4 : public au::Day<AreaTuple>
{
public:
    Day4() : Day(2022, 4) {}

    int test_one(const vector<AreaTuple> &input)
    {
        unsigned int cnt = 0;
        for (AreaTuple at : input)
            if (at.contains())
                cnt++;

        return cnt;
    }

    int part_one(const vector<AreaTuple> &input)
    {
        unsigned int cnt = 0;
        for (AreaTuple at : input)
            if (at.contains())
                cnt++;

        return cnt;
    }

    int test_two(const vector<AreaTuple> &input)
    {
        unsigned int cnt = 0;
        for (AreaTuple at : input)
            if (at.overlap())
                cnt++;

        return cnt;
    }

    int part_two(const vector<AreaTuple> &input)
    {
        unsigned int cnt = 0;
        for (AreaTuple at : input)
            if (at.overlap())
                cnt++;

        return cnt;
    }
};

int main()
{
    Day4().run_all(true);

    return 0;
}
